Update to Python 3.8

gondiaz opened the pull request:
This PR is intended to update the IC conda and libraries to allow working in Python ~~3.9~~ 3.8.

The branch has been tested on the following systems (we will update the list once more machines are tested):

✅ MacOS Monterey (12.0.1)
✅ Ubuntu 20.04 (new system for github actions)
~~❌ Ubuntu 18.04 (old system for github actions) --> failing tests~~
✅   Ubuntu 16.04 (gpu1next)
✅   Red Hat Enterprise Linux Server 7.5 (Maipo) (CESGA)
✅ CentOS Linux 7 (majorana1)
✅ CentOS Linux 7 (frontend1demo)
❌ CentOS Linux 7 (PowerPC) (argudell) --> failing tests (conda has to be installed appart, ppc64el arch)
❌ Scientific Linux 6.6 (neutrinos1) --> failing conda installation


~~Failing tests may be related with system incompatibility with newer scipy version, since they fail for tests using scipy functions producing nans, in particular in IC random-sampling and deconvolution functions.~~

The Ubuntu version of github actions has been updated to Ubuntu 20.04 from Ubuntu 18.04 to allow tests to pass.

We would appreciate some input from other machines.

gondiaz committed:
Update IC libraries

gondiaz committed:
Resolve most failing tests from empty tables

This commit shows most failing tests are due to past tests files where
empty tables are saved as floats. This commits avoid to assert on
empty tables.

gondiaz committed:
Sort by row index in testing function

gondiaz committed:
Solve psf functions tests

gondiaz committed:
Update to python version 3.8

gondiaz committed:
Update MC exact tests files

gondiaz committed:
Revert: Resolve most failing tests from empty tables

This reverts commit 0a5beecf0b35a94e6d0812dc524972b571c7f9ac.

gondiaz committed:
Update conda version

gondiaz committed:
Clean up np.type deprecation warnings

Some np.type names have been deprecated since numpy 1.20.0
release, the Python builtin should be used instead. See https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations

jmbenlloch committed:
Update ubuntu version in github action

Update ubuntu and python versions in calibration check

jacg commented on invisible_cities/reco/psf_functions_test.py:
```diff
@@ -55,6 +55,7 @@ def test_add_empty_sensors_and_normalize_q_file(ICDATADIR):
                                  database = load_db.DataSiPM('new', 0))
     hdst_processed.reset_index(inplace=True, drop=True)
     hdst_psf       = pd.read_hdf(PATH_TEST)
+    hdst_psf       = hdst_psf.astype(hdst_processed.dtypes) # since pandas version update
```
Minor point, but I'd mention the old and new version in the comment about the pandas version update.

And again when the same comment appears on line 72

gondiaz committed:
Update README

gondiaz committed:
Downgrade numpy to solve buffy tests

gondiaz committed:
Remove deadline in variable time tests

This avoids failing tests on some machines due to larger timing than
the default deadline

Simplify decorator syntax

Change decorator ordering

gonzaponte committed:
Bump pytables to 3.7.0

v3.6.1 had a annoying warning that appeared 19k times in our tests

gonzaponte committed:
Update conda env tag

gonzaponte committed:
Update dataframe comparison functions

Due to a change in the pandas API

gonzaponte committed:
Update calls to dataframe comparison functions

gonzaponte committed:
Filter warnings

gonzaponte committed:
Add regex=True to silence warning

Accounts for a future change in the API

gonzaponte commented on invisible_cities/detsim/light_tables_c_test.py:
```diff
@@ -34,6 +34,7 @@ def test_LT_SiPM_optional_arguments(get_dfs):
 @given(xs=floats(min_value=-500, max_value=500),
        ys=floats(min_value=-500, max_value=500),
        sipm_indx=integers(min_value=0, max_value=1500))
+@settings(deadline=None, max_examples=100)
```
can we define `few_examples = settings(deadline=None, max_examples=100)`? Also I think it makes sense for `given` to be the closest decorator to the function definition.
Also, how bad was the deadline issue?

gondiaz replied:
If I remember correctly, the issue was not serious but due to a change in the default deadline of pytest in newer versions. Also, such failing tests depended on the machine, one may be faster than others in carrying the computation, I guess. I decorated all the failing tests of the machines I tested, then it may be more tests with this deadline issue in some particular machines.

gonzaponte reviewed (commented):
What is this thing of going from `np.type` to `type`? do you remember? If so, can you reword the commit message, so it includes the explanation?

gonzaponte reviewed (approved):
Important update to a more recent python version and this repository's dependencies.
Upgrade-related issues addressed clearly and warnings silenced.
Hopefully, the next upgrade won't take this long :)
Good job.
